design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,195 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to immediately propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out DaemonSets.
+
+
+## Goals
+
+-   Provide a way to propagate changes from a running VM object to its corresponding VMI immediately without interrupting the workload.
+    
+-   Maintain backwards compatibility
+    
+
+## Non Goals
+
+-   Successfully apply all the changes to the Virtual Machine Instance Object. Not all fields can be propagated to the guest immediately. Some fields should remain staged, such as those that affect the guest ABI (e.g. machine type)
+    
+-   Make LiveUpdate the default. We defer this decision.
+    
+
+## Definition of Users
+
+A user - a VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a user I want changes that I applied to the VM object to propagate to the VMI immediately if possible.
+
+As a user, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+A new field called rolloutStrategy will be introduced in the VM spec. This field will control the behavior of the VM object. Two possible values will be available:
+
+1.  Setting RolloutStrategy to `LiveUpdate` means all changes to the VM object will be immediately applied to the VMI.
+    
+
+A condition called `RestartRequired` will be added to the VM if such changes are not possible to propagate,
+
+2.  Setting it to `Stage` would mean the current familiar behavior remains - all changes to the VM object will be staged and applied after a restart.
+    
+
+The same rolloutStrategy field will be added to the KubeVirt CR and will be the default cluster-wide.
+
+
+Interaction with existing `liveUpdateFeatures` field:
+
+1.  When `LiveUpdate` is used `liveUpdateFeatures` don’t need to be explicitly defined because they are enabled by default
+    
+2.  When `Stage` is used, `liveUpdateFeatures` will be ignored, except for special cases related to upgrades, see below.
+
+The above effectively means that the `LiveUpdateFeatures` field will be removed.
+
+The default value of the `rolloutStrategy` field if not specified will be `Stage`.
+
+If an existing VM (on upgrade) has the `LiveUpdateFeatures` field specified the `rolloutStrategy` field will be `LiveUpdate` and its `LiveUpdateFeatures` field will be removed.
+
+
+The `rolloutStrategy` field will be usable only if the `VMLiveUpdateFeatures` feature gate is enabled.
```

Comment:
If this was only a cluster-level option (i.e. in the KV CR), could we just use the FG instead of adding a new field?

Reply:
For me, it would take a similar approach to the [DaemonSets](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy) update strategy: `OnDelete` and `RollingUpdate` or Deployments `strategy`. 

I think that if we only provide a cluster-level option, users will lose the flexibility to select different behavior for certain VMs.

Reply:
Ryan's VMPool example/request is along those lines as well.

Reply:
Yea, the FG is not a fit to express how things should behave, only as means to enable/disable a feature and as part of its introduction lifecycle.